File: application.py
# run python3 application.py NOT flask run
# export FLASK_ENV=development
# export FLASK_ENV=production
# export DATABASE_URL=sqlite:////tmp/test.db

import os
from flask import Flask, escape, request, render_template
from flask_socketio import SocketIO, emit
from models import *
from flask import current_app as app
from get_data_2 import *
from api_2 import *
import cProfile

# get data on startup
with app.app_context():
    socketio = SocketIO(app)
    # print("getting data...")
    # get_and_save_data('Spring 2020')
    # print("got data")

@app.route('/')
def index():
    return render_template('index.html')

@socketio.on("update data")
def update_data():
    t = time.time()
    get_and_save_data('Spring 2020')
    print(time.time() - t)
    emit('updated data')

@socketio.on("search")
def search(data):

    print('serch term:', data['term'])
    search_results = get_search_results(data['term'])
    search_results_json = [{
        'course_num': x.course_num,
        'title': x.title,
        'desc_long': x.desc_long,
        'id': x.id
    } for x in search_results]

    # names = ['MATH-0166', 'PHIL-0192', 'MATH-0070']
    # t = time.time()
    # courses = [get_search_results(x)[0] for x in names]
    # print(time.time() - t)
    # t = time.time()
    # pg = PeriodGroup([course_object_to_period_group(x, accepted_statuses=('O','W','C')) for x in courses], 'and')
    # print(time.time() - t)
    # t = time.time()
    # schedules = pg.evaluate()
    # print(time.time() - t)
    # print(schedules)

    # cProfile.run("[get_search_results(x)[0] for x in ['MATH-0166', 'PHIL-0192', 'MATH-0070']]", sort='cumulative')

    



    emit('results', search_results_json)


if __name__ == '__main__':
    socketio.run(app)

File: application.js
const express = require('express');
const app = express();
app.use(express.static('static'));
const http = require('http').createServer(app);

const assert = require('assert');
const fs = require('fs');

const models = require('./models');
const api = require('./api');
const get_data = require('./get_data');
const course_scheduler = require('./course_scheduler');
const evaluation_timing_test = require('./some_tests').evaluation_timing_test;


function time() {
    return Date.now();
}

app.get('/', function(req, res) {
  res.sendFile(__dirname + '/templates/index.html');
});

app.get('/get_schedules', function(req, res) {
  console.log("Got request to gnerate schedules");

  const ids = req.query.ids.map(id => parseInt(id, 10));
  const courses = [];
  for (const id of ids)
    courses.push(models.courses[id]);
  const schedules = get_schedules(courses, ['O', 'C', 'W']);
  let t = time();
  const schedules_list = [];
  for (const s of schedules) {
    schedules_list.push(s);
  }
  console.log("time taken:", time() - t)
  console.log("number of schedules:", schedules_list.length);
  if (schedules_list.length < 100)
    console.log(schedules_list);









  
  // const [courses, schedules] = get_schedules(ids, ['O', 'C', 'W']);
  // let t = time();
  // const schedules_list = [];
  // for (const s of schedules) {
  //   schedules_list.push(s);
  // }
  // console.log("time taken:", time() - t)
  // console.log("number of schedules:", schedules_list.length);
  // console.log(schedules_list);


  

  // const sections_by_id_to_send = {};
  // for (const course of courses)
  //   for (const section of course.sections)
  //     sections_by_id_to_send[section.id] = section;
  // console.log("sections string length:", JSON.stringify(sections_by_id_to_send).length);
  // console.log("schedules string length:", JSON.stringify(schedules_list).length);


  // // get courses
  // const courses = [];
  // for (const id of ids)
  //   courses.push(models.courses[id]);
  // // redo ids to start from 0
  // const id_to_id = {};
  // const sections_by_id_to_send = {};
  // let curr_id = 0;
  // for (const course of courses) {
  //   for (const section of course.sections) {
  //     id_to_id[section.id] = curr_id++;
  //     sections_by_id_to_send[curr_id] = section;
  //   }
  // }
  // const schedules = get_schedules(courses, ['O', 'C', 'W']);
  // let t = time();
  // const schedules_list = [];
  // var nums_iterable = function*() {
  //   for (const s of get_schedules(courses, ['O', 'C', 'W']))
  //     for (const subschedule of s)
  //       for (const num of subschedule)
  //         yield 0;
  // }();
  // var nums = new Uint16Array(nums_iterable);

  // var nums = new Uint16Array(138489510);
  // var n = 0;
  // for (const s of get_schedules(courses, ['O', 'C', 'W'])) {
  //   for (const subschedule of s) {
  //     for (const num of subschedule) {
  //       nums[++n] = id_to_id[num];
  //     }
  //   }
  // }
  // res.send(nums);

  

  // console.log("sections string length:", JSON.stringify(sections_by_id_to_send).length);
  // console.log("schedules string length:", schedules_string.length);
  // if (schedules_string.length > 400)
  //   console.log(schedules_string.slice(0, 400));
  // else
  //   console.log(schedules_string);


  // let n = 0;
  // for (const s of get_schedules(courses, ['O', 'C', 'W']))
  //   for (const subschedule of s)
  //     for (const num of subschedule)
  //       n++;
  // console.log("IS HOW MANY:", n);


  // console.log(schedules);
  // console.log(JSON.stringify(schedules));
  // let t = time();
  // const schedules_ids = schedules.map(schedule => schedule.map(sections => sections.map(section => section.id)));
  // console.log(time() - t);
  // console.log(schedules_ids);

  res.send({});
});

app.get('/search/:term', function(req, res) {
  console.log(`search term: ${req.params.term}`);
  const search_results = api.get_search_results(req.params.term);
  const search_results_json = search_results.map(course => ({
      course_num: course.course_num,
      title: course.title,
      desc_long: course.desc_long,
      id: course.id
  }));
  res.send(search_results_json);
});

app.get('/updatedata', function(req, res) {
  console.log('going to update data...');
  get_data.get_and_save_data('Spring 2020', function() {
    res.send('updated data');
  });
});

http.listen(3000, function() {
  console.log('listening on *:3000');
});

function get_schedules(courses, accepted_statuses) {
  const pg = new course_scheduler.PeriodGroup(
    courses.map(course => api.course_object_to_period_group(course, true, accepted_statuses, cache=false, give_ids=true)),
    'and', merge=false, cache=false
  );
  return pg.evaluate();
}

// function get_schedules(course_ids, accepted_statuses) {
//   // const courses = models.courses.filter(course => course_ids.includes(course.id));
//   const courses = [];
//   for (const id of course_ids)
//     courses.push(models.courses[id]);
//   const pg = new course_scheduler.PeriodGroup(
//     courses.map(course => api.course_object_to_period_group(course, true, accepted_statuses, cache=false, give_ids=true)),
//     'and', merge=false, cache=false
//   );
//   return [courses, pg.evaluate()];
// }
